File: port/python/src/muonledger/commands/balance.py
"""Balance command -- produces a balance report from a journal.

Ported from ledger's ``output.cc`` (``format_accounts``) and the balance
report logic.  Given a :class:`Journal`, accumulates posting amounts into
their accounts and renders a formatted balance report.

The output format mirrors ledger's default balance format::

               $100.00  Assets:Bank:Checking
               $-50.00  Expenses:Food
    --------------------
                $50.00

"""

from __future__ import annotations

import argparse
from datetime import date, datetime
from typing import Optional

from muonledger.account import Account
from muonledger.amount import Amount
from muonledger.balance import Balance
from muonledger.journal import Journal

__all__ = ["balance_command"]

# Column width for amounts (right-aligned within this width).
AMOUNT_WIDTH = 20
# Separator line width matches amount column.
SEPARATOR = "-" * AMOUNT_WIDTH


def _parse_date_arg(text: str) -> date:
    """Parse a date argument in YYYY-MM-DD or YYYY/MM/DD format."""
    for fmt in ("%Y-%m-%d", "%Y/%m/%d"):
        try:
            return datetime.strptime(text, fmt).date()
        except ValueError:
            continue
    raise ValueError(f"Cannot parse date: {text!r}")


def _parse_args(args: list[str]) -> argparse.Namespace:
    """Parse balance command arguments."""
    parser = argparse.ArgumentParser(prog="balance", add_help=False)
    parser.add_argument("--flat", action="store_true", default=False)
    parser.add_argument("--no-total", action="store_true", default=False)
    parser.add_argument("-n", "--collapse", action="store_true", default=False,
                        dest="collapse")
    parser.add_argument("--empty", "-E", action="store_true", default=False)
    parser.add_argument("--depth", type=int, default=0)
    parser.add_argument("--begin", type=_parse_date_arg, default=None)
    parser.add_argument("--end", type=_parse_date_arg, default=None)
    parser.add_argument("patterns", nargs="*")
    return parser.parse_args(args)


def _accumulate_balances(journal: Journal, begin: Optional[date] = None,
                         end: Optional[date] = None) -> dict[str, Balance]:
    """Walk all transactions and accumulate per-account balances.

    Returns a dict mapping full account name to its Balance.
    """
    balances: dict[str, Balance] = {}
    for xact in journal.xacts:
        # Date filtering: --begin means >= date, --end means < date
        if begin is not None and xact.date is not None:
            if xact.date < begin:
                continue
        if end is not None and xact.date is not None:
            if xact.date >= end:
                continue
        for post in xact.posts:
            if post.amount is None or post.amount.is_null():
                continue
            acct = post.account
            if acct is None:
                continue
            name = acct.fullname
            if name not in balances:
                balances[name] = Balance()
            balances[name].add(post.amount)
    return balances


def _roll_up_to_parents(
    balances: dict[str, Balance],
) -> dict[str, Balance]:
    """Roll up leaf balances into all ancestor accounts.

    Returns a new dict containing every account (leaf and parent) with its
    rolled-up total.
    """
    rolled: dict[str, Balance] = {}
    for name, bal in balances.items():
        parts = name.split(":")
        for i in range(1, len(parts) + 1):
            ancestor = ":".join(parts[:i])
            if ancestor not in rolled:
                rolled[ancestor] = Balance()
            rolled[ancestor].add(bal)
    return rolled


def _get_children(name: str, all_names: set[str]) -> list[str]:
    """Return immediate children of *name* from the set of all account names."""
    prefix = name + ":"
    children = []
    for n in all_names:
        if not n.startswith(prefix):
            continue
        rest = n[len(prefix):]
        if ":" not in rest:
            children.append(n)
    return sorted(children)


def _account_known(name: str, journal: Journal) -> bool:
    """Return True if the account exists in the journal's account tree."""
    return journal.master.find_account(name, auto_create=False) is not None


def _collect_tree_accounts(
    rolled: dict[str, Balance],
    leaf_balances: dict[str, Balance],
    show_empty: bool,
    depth: int,
    journal: Journal,
) -> list[tuple[str, str, Balance]]:
    """Determine which accounts to show in tree mode.

    Returns a list of (display_name, full_name, balance) tuples.

    Ledger's tree mode collapses single-child chains: if an account has
    exactly one visible child and no direct postings, the account is not
    shown on its own line; instead, its child absorbs it and displays
    with a collapsed name (e.g. ``Equity:Opening balances``).

    The display_name for child accounts whose parents are already
    displayed uses only the leaf segment of the name.
    """
    all_names = set(rolled.keys())
    top_level = sorted(n for n in all_names if ":" not in n)

    result: list[tuple[str, str, Balance]] = []

    def _visible_children(name: str) -> list[str]:
        children = _get_children(name, all_names)
        if not show_empty:
            children = [c for c in children
                        if rolled.get(c, Balance()).is_nonzero()]
        return children

    def _has_direct(name: str) -> bool:
        """Does this account have direct postings with non-zero balance?"""
        return name in leaf_balances and leaf_balances[name].is_nonzero()

    def _has_direct_or_known(name: str) -> bool:
        """Check if account has direct postings (appears in leaf_balances).

        This determines whether an account can be collapsed. An account
        with direct postings should NOT be collapsed even if it has only
        one child, because collapsing would hide the account's own entry.
        """
        return name in leaf_balances

    def _walk(name: str, current_depth: int, collapse_prefix: str,
              indent_depth: int) -> None:
        """Walk the account tree depth-first.

        *collapse_prefix* accumulates collapsed parent names. When a
        parent with a single child and no direct postings is encountered,
        its name is prepended to the child's display name via this
        parameter.

        *indent_depth* tracks the visual indentation level (number of
        parent accounts already displayed above this one).
        """
        if depth > 0 and current_depth >= depth:
            return

        bal = rolled.get(name, Balance())
        children = _visible_children(name)

        # Build the display name for this node.
        leaf_segment = name.rsplit(":", 1)[-1]
        if collapse_prefix:
            display = collapse_prefix + ":" + leaf_segment
        else:
            display = leaf_segment

        # Collapse: if exactly one child and no direct postings,
        # pass our display name down to the child.
        if len(children) == 1 and not _has_direct_or_known(name):
            _walk(children[0], current_depth, display, indent_depth)
            return

        # Should we display this account?
        should_show = False
        if bal.is_nonzero():
            should_show = True
        elif show_empty and _has_direct_or_known(name):
            should_show = True

        if not should_show and not children:
            return

        if should_show:
            # Add indentation: 2 spaces per indent level
            indented_display = "  " * indent_depth + display
            result.append((indented_display, name, bal))

        for child in children:
            child_prefix = "" if should_show else display
            _walk(child, current_depth + 1, child_prefix,
                  indent_depth + 1 if should_show else indent_depth)

    for top in top_level:
        _walk(top, 0, "", 0)

    return result


def _format_amount_lines(bal: Balance) -> list[str]:
    """Format a Balance into right-aligned amount strings.

    Each commodity gets its own line (multi-commodity support).
    If the balance is empty/zero, returns a single "0" line.
    """
    if bal.is_empty():
        return [str(0).rjust(AMOUNT_WIDTH)]

    lines = []
    for amt in bal:  # iterates in sorted commodity order
        lines.append(str(amt).rjust(AMOUNT_WIDTH))

    if not lines:
        lines.append(str(0).rjust(AMOUNT_WIDTH))

    return lines


def balance_command(
    journal: Journal,
    args: Optional[list[str]] = None,
) -> str:
    """Produce a balance report from *journal*.

    Parameters
    ----------
    journal:
        The journal containing transactions to report on.
    args:
        Command-line style arguments (e.g. ``["--flat", "Assets"]``).

    Returns
    -------
    str
        The formatted balance report.
    """
    if args is None:
        args = []
    opts = _parse_args(args)
    no_total = opts.no_total
    patterns = opts.patterns

    # -n / --collapse is equivalent to --depth 1 in tree mode.
    # In flat mode, -n produces no output (collapse is meaningless for flat).
    effective_depth = opts.depth
    if opts.collapse and not opts.flat and effective_depth == 0:
        effective_depth = 1

    # Step 1: Accumulate per-account (leaf) balances.
    leaf_balances = _accumulate_balances(journal, begin=opts.begin, end=opts.end)

    # Step 2: Roll up balances to parents.
    rolled = _roll_up_to_parents(leaf_balances)

    # Step 3: Apply depth limiting.
    if effective_depth > 0:
        rolled = _apply_depth(rolled, effective_depth)
        leaf_balances_depth: dict[str, Balance] = {}
        for name, bal in leaf_balances.items():
            parts = name.split(":")
            if len(parts) > effective_depth:
                truncated = ":".join(parts[: effective_depth])
            else:
                truncated = name
            if truncated not in leaf_balances_depth:
                leaf_balances_depth[truncated] = Balance()
            leaf_balances_depth[truncated].add(bal)
        leaf_balances = leaf_balances_depth

    # Step 4: Determine which accounts to display.
    if opts.flat and opts.collapse:
        # -n/--collapse with --flat produces no output in C++ ledger.
        accounts = []
    elif opts.flat:
        accounts = _flat_accounts(
            leaf_balances, rolled, patterns, opts.empty, effective_depth
        )
    else:
        accounts = _tree_accounts(
            leaf_balances, rolled, patterns, opts.empty, effective_depth, journal
        )

    # Step 5: Render.
    lines: list[str] = []

    for display_name, full_name, bal in accounts:
        amt_lines = _format_amount_lines(bal)
        lines.append(f"{amt_lines[0]}  {display_name}")
        for extra in amt_lines[1:]:
            lines.append(extra)

    # Compute grand total from leaf balances (avoids double-counting in
    # tree mode where parent and child are both displayed).
    grand_total = Balance()
    for bal in leaf_balances.values():
        grand_total.add(bal)

    # If patterns are active, compute total only from matching accounts.
    if patterns:
        grand_total = Balance()
        for display_name, full_name, bal in accounts:
            # In flat mode, each entry is a leaf -- safe to sum.
            # In tree mode, we only sum entries that have no displayed
            # children to avoid double-counting.
            if opts.flat:
                grand_total.add(bal)
            else:
                # Use leaf balances for matching accounts.
                if full_name in leaf_balances:
                    grand_total.add(leaf_balances[full_name])

    # Total line.
    if not no_total and len(accounts) > 0:
        lines.append(SEPARATOR)
        if grand_total.is_empty() or grand_total.is_zero():
            lines.append(str(0).rjust(AMOUNT_WIDTH))
        else:
            total_lines = _format_amount_lines(grand_total)
            for tl in total_lines:
                lines.append(tl)

    if not lines:
        return ""

    return "\n".join(lines) + "\n"


def _apply_depth(
    rolled: dict[str, Balance], depth: int
) -> dict[str, Balance]:
    """Truncate account names to *depth* segments and re-merge."""
    result: dict[str, Balance] = {}
    for name, bal in rolled.items():
        parts = name.split(":")
        if len(parts) <= depth:
            result[name] = Balance(bal)
    return result


def _flat_accounts(
    leaf_balances: dict[str, Balance],
    rolled: dict[str, Balance],
    patterns: list[str],
    show_empty: bool,
    depth: int,
) -> list[tuple[str, str, Balance]]:
    """Collect accounts for flat display."""
    result: list[tuple[str, str, Balance]] = []
    for name in sorted(leaf_balances.keys()):
        bal = leaf_balances[name]
        if not show_empty and not bal.is_nonzero():
            continue
        if patterns and not _matches(name, patterns):
            continue
        if depth > 0 and len(name.split(":")) > depth:
            continue
        result.append((name, name, bal))
    return result


def _tree_accounts(
    leaf_balances: dict[str, Balance],
    rolled: dict[str, Balance],
    patterns: list[str],
    show_empty: bool,
    depth: int,
    journal: Journal,
) -> list[tuple[str, str, Balance]]:
    """Collect accounts for tree (default) display."""
    entries = _collect_tree_accounts(
        rolled, leaf_balances, show_empty, depth, journal
    )

    if patterns:
        matching_full = set()
        for _, full, _ in entries:
            if _matches(full, patterns):
                matching_full.add(full)
        ancestor_names = set()
        for fn in matching_full:
            parts = fn.split(":")
            for i in range(1, len(parts)):
                ancestor_names.add(":".join(parts[:i]))

        filtered = []
        for display, full, bal in entries:
            if full in matching_full or full in ancestor_names:
                filtered.append((display, full, bal))
        entries = filtered

    return entries


def _matches(name: str, patterns: list[str]) -> bool:
    """Simple substring match: True if *name* contains any pattern."""
    name_lower = name.lower()
    return any(p.lower() in name_lower for p in patterns)

File: port/rust/src/commands/balance.rs
//! Balance command -- produces a balance report from a journal.
//!
//! Ported from the Python reference implementation. Given a [`Journal`],
//! accumulates posting amounts into their accounts and renders a formatted
//! balance report.
//!
//! The output format mirrors ledger's default balance format:
//!
//! ```text
//!            $100.00  Assets:Bank:Checking
//!            $-50.00  Expenses:Food
//! --------------------
//!             $50.00
//! ```

use std::collections::{BTreeMap, HashSet};

use crate::balance::Balance;
use crate::journal::Journal;

/// Column width for amounts (right-aligned within this width).
const AMOUNT_WIDTH: usize = 20;

/// Separator line width matches amount column.
fn separator() -> String {
    "-".repeat(AMOUNT_WIDTH)
}

/// Options for the balance command.
#[derive(Debug, Clone)]
pub struct BalanceOptions {
    /// Show flat (non-hierarchical) output.
    pub flat: bool,
    /// Suppress the total line.
    pub no_total: bool,
    /// Show accounts with zero balances.
    pub show_empty: bool,
    /// Limit display depth (0 = unlimited).
    pub depth: usize,
    /// Account name filter patterns (substring match, case-insensitive).
    pub patterns: Vec<String>,
}

impl Default for BalanceOptions {
    fn default() -> Self {
        Self {
            flat: false,
            no_total: false,
            show_empty: false,
            depth: 0,
            patterns: Vec::new(),
        }
    }
}

/// Accumulate per-account (leaf) balances from all transactions.
fn accumulate_balances(journal: &Journal) -> BTreeMap<String, Balance> {
    let mut balances: BTreeMap<String, Balance> = BTreeMap::new();
    for xact in &journal.xacts {
        for post in &xact.posts {
            let amt = match &post.amount {
                Some(a) if !a.is_null() => a,
                _ => continue,
            };
            let account_id = match post.account_id {
                Some(id) => id,
                None => continue,
            };
            let name = journal.accounts.fullname(account_id);
            if name.is_empty() {
                continue;
            }
            let bal = balances.entry(name).or_insert_with(Balance::new);
            bal.add_amount(amt).ok();
        }
    }
    balances
}

/// Roll up leaf balances into all ancestor accounts.
fn roll_up_to_parents(balances: &BTreeMap<String, Balance>) -> BTreeMap<String, Balance> {
    let mut rolled: BTreeMap<String, Balance> = BTreeMap::new();
    for (name, bal) in balances {
        let parts: Vec<&str> = name.split(':').collect();
        for i in 1..=parts.len() {
            let ancestor = parts[..i].join(":");
            let entry = rolled.entry(ancestor).or_insert_with(Balance::new);
            entry.add_balance(bal);
        }
    }
    rolled
}

/// Apply depth limiting: keep only accounts with <= depth segments.
fn apply_depth(rolled: &BTreeMap<String, Balance>, depth: usize) -> BTreeMap<String, Balance> {
    let mut result: BTreeMap<String, Balance> = BTreeMap::new();
    for (name, bal) in rolled {
        let segments = name.split(':').count();
        if segments <= depth {
            result.insert(name.clone(), bal.clone());
        }
    }
    result
}

/// Simple substring match (case-insensitive).
fn matches_pattern(name: &str, patterns: &[String]) -> bool {
    if patterns.is_empty() {
        return true;
    }
    let lower = name.to_lowercase();
    patterns.iter().any(|p| lower.contains(&p.to_lowercase()))
}

/// Return immediate children of `name` from the set of all account names.
fn get_children(name: &str, all_names: &HashSet<String>) -> Vec<String> {
    let prefix = format!("{}:", name);
    let mut children = Vec::new();
    for n in all_names {
        if !n.starts_with(&prefix) {
            continue;
        }
        let rest = &n[prefix.len()..];
        if !rest.contains(':') {
            children.push(n.clone());
        }
    }
    children.sort();
    children
}

/// Account entry for display: (display_name, full_name, balance).
type AccountEntry = (String, String, Balance);

/// Collect accounts for tree (hierarchical) display.
fn collect_tree_accounts(
    rolled: &BTreeMap<String, Balance>,
    leaf_balances: &BTreeMap<String, Balance>,
    show_empty: bool,
    depth: usize,
) -> Vec<AccountEntry> {
    let all_names: HashSet<String> = rolled.keys().cloned().collect();
    let top_level: Vec<String> = {
        let mut tops: Vec<String> = all_names
            .iter()
            .filter(|n| !n.contains(':'))
            .cloned()
            .collect();
        tops.sort();
        tops
    };

    let mut result: Vec<AccountEntry> = Vec::new();

    fn visible_children(
        name: &str,
        all_names: &HashSet<String>,
        rolled: &BTreeMap<String, Balance>,
        show_empty: bool,
    ) -> Vec<String> {
        let children = get_children(name, all_names);
        if !show_empty {
            children
                .into_iter()
                .filter(|c| {
                    rolled
                        .get(c)
                        .map_or(false, |b| b.is_nonzero())
                })
                .collect()
        } else {
            children
        }
    }

    fn has_direct_or_known(name: &str, leaf_balances: &BTreeMap<String, Balance>) -> bool {
        leaf_balances.contains_key(name)
    }

    fn walk(
        name: &str,
        current_depth: usize,
        collapse_prefix: &str,
        indent_depth: usize,
        depth_limit: usize,
        all_names: &HashSet<String>,
        rolled: &BTreeMap<String, Balance>,
        leaf_balances: &BTreeMap<String, Balance>,
        show_empty: bool,
        result: &mut Vec<AccountEntry>,
    ) {
        if depth_limit > 0 && current_depth >= depth_limit {
            return;
        }

        let bal = rolled.get(name).cloned().unwrap_or_else(Balance::new);
        let children = visible_children(name, all_names, rolled, show_empty);

        // Build display name
        let leaf_segment = name.rsplit(':').next().unwrap_or(name);
        let display = if !collapse_prefix.is_empty() {
            format!("{}:{}", collapse_prefix, leaf_segment)
        } else {
            leaf_segment.to_string()
        };

        // Collapse: single child, no direct postings
        if children.len() == 1 && !has_direct_or_known(name, leaf_balances) {
            walk(
                &children[0],
                current_depth,
                &display,
                indent_depth,
                depth_limit,
                all_names,
                rolled,
                leaf_balances,
                show_empty,
                result,
            );
            return;
        }

        // Should we display this account?
        let mut should_show = false;
        if bal.is_nonzero() {
            should_show = true;
        } else if show_empty && has_direct_or_known(name, leaf_balances) {
            should_show = true;
        }

        if !should_show && children.is_empty() {
            return;
        }

        if should_show {
            let indented = format!("{}{}", "  ".repeat(indent_depth), display);
            result.push((indented, name.to_string(), bal));
        }

        for child in &children {
            walk(
                child,
                current_depth + 1,
                "",
                indent_depth + if should_show { 1 } else { 0 },
                depth_limit,
                all_names,
                rolled,
                leaf_balances,
                show_empty,
                result,
            );
        }
    }

    for top in &top_level {
        walk(
            top,
            0,
            "",
            0,
            depth,
            &all_names,
            rolled,
            leaf_balances,
            show_empty,
            &mut result,
        );
    }

    result
}

/// Collect accounts for flat display.
fn flat_accounts(
    leaf_balances: &BTreeMap<String, Balance>,
    patterns: &[String],
    show_empty: bool,
    depth: usize,
) -> Vec<AccountEntry> {
    let mut result: Vec<AccountEntry> = Vec::new();
    for (name, bal) in leaf_balances {
        if !show_empty && !bal.is_nonzero() {
            continue;
        }
        if !patterns.is_empty() && !matches_pattern(name, patterns) {
            continue;
        }
        if depth > 0 && name.split(':').count() > depth {
            continue;
        }
        result.push((name.clone(), name.clone(), bal.clone()));
    }
    result
}

/// Format a Balance into right-aligned amount strings.
fn format_amount_lines(bal: &Balance) -> Vec<String> {
    if bal.is_empty() {
        return vec![format!("{:>width$}", 0, width = AMOUNT_WIDTH)];
    }

    let mut lines: Vec<String> = Vec::new();
    for amt in bal.iter() {
        lines.push(format!("{:>width$}", amt.to_string(), width = AMOUNT_WIDTH));
    }

    if lines.is_empty() {
        lines.push(format!("{:>width$}", 0, width = AMOUNT_WIDTH));
    }

    lines
}

/// Produce a balance report from a journal.
///
/// # Arguments
///
/// * `journal` - The journal containing transactions to report on.
/// * `opts` - Options controlling the report format.
///
/// # Returns
///
/// The formatted balance report as a string.
pub fn balance_command(journal: &Journal, opts: &BalanceOptions) -> String {
    let effective_depth = opts.depth;

    // Step 1: Accumulate per-account (leaf) balances.
    let mut leaf_balances = accumulate_balances(journal);

    // Step 2: Roll up balances to parents.
    let mut rolled = roll_up_to_parents(&leaf_balances);

    // Step 3: Apply depth limiting.
    if effective_depth > 0 {
        rolled = apply_depth(&rolled, effective_depth);
        let mut depth_leaves: BTreeMap<String, Balance> = BTreeMap::new();
        for (name, bal) in &leaf_balances {
            let parts: Vec<&str> = name.split(':').collect();
            let truncated = if parts.len() > effective_depth {
                parts[..effective_depth].join(":")
            } else {
                name.clone()
            };
            let entry = depth_leaves.entry(truncated).or_insert_with(Balance::new);
            entry.add_balance(bal);
        }
        leaf_balances = depth_leaves;
    }

    // Step 4: Determine which accounts to display.
    let accounts = if opts.flat {
        flat_accounts(&leaf_balances, &opts.patterns, opts.show_empty, effective_depth)
    } else {
        let entries = collect_tree_accounts(&rolled, &leaf_balances, opts.show_empty, effective_depth);
        if !opts.patterns.is_empty() {
            filter_tree_by_patterns(&entries, &opts.patterns)
        } else {
            entries
        }
    };

    // Step 5: Render.
    let mut lines: Vec<String> = Vec::new();

    for (display_name, _full_name, bal) in &accounts {
        let amt_lines = format_amount_lines(bal);
        lines.push(format!("{}  {}", amt_lines[0], display_name));
        for extra in &amt_lines[1..] {
            lines.push(extra.clone());
        }
    }

    // Compute grand total from leaf balances.
    let mut grand_total = Balance::new();
    if opts.patterns.is_empty() {
        for bal in leaf_balances.values() {
            grand_total.add_balance(bal);
        }
    } else {
        // When patterns are active, sum only matching accounts.
        if opts.flat {
            for (_display, _full, bal) in &accounts {
                grand_total.add_balance(bal);
            }
        } else {
            for (_display, full, _bal) in &accounts {
                if let Some(leaf_bal) = leaf_balances.get(full) {
                    grand_total.add_balance(leaf_bal);
                }
            }
        }
    }

    // Total line.
    if !opts.no_total && !accounts.is_empty() {
        lines.push(separator());
        if grand_total.is_empty() || grand_total.is_zero() {
            lines.push(format!("{:>width$}", 0, width = AMOUNT_WIDTH));
        } else {
            let total_lines = format_amount_lines(&grand_total);
            for tl in &total_lines {
                lines.push(tl.clone());
            }
        }
    }

    if lines.is_empty() {
        return String::new();
    }

    format!("{}\n", lines.join("\n"))
}

/// Filter tree entries by patterns, keeping ancestors of matching accounts.
fn filter_tree_by_patterns(entries: &[AccountEntry], patterns: &[String]) -> Vec<AccountEntry> {
    let mut matching_full: HashSet<String> = HashSet::new();
    for (_display, full, _bal) in entries {
        if matches_pattern(full, patterns) {
            matching_full.insert(full.clone());
        }
    }

    let mut ancestor_names: HashSet<String> = HashSet::new();
    for full in &matching_full {
        let parts: Vec<&str> = full.split(':').collect();
        for i in 1..parts.len() {
            ancestor_names.insert(parts[..i].join(":"));
        }
    }

    entries
        .iter()
        .filter(|(_display, full, _bal)| {
            matching_full.contains(full) || ancestor_names.contains(full)
        })
        .cloned()
        .collect()
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::amount::Amount;
    use crate::journal::Journal;
    use crate::parser::TextualParser;
    use crate::post::Post;
    use crate::xact::Transaction;

    fn parse_journal(text: &str) -> Journal {
        let mut journal = Journal::new();
        let parser = TextualParser::new();
        parser.parse_string(text, &mut journal).unwrap();
        journal
    }

    // ---- basic tests -------------------------------------------------------

    #[test]
    fn balance_empty_journal() {
        let journal = Journal::new();
        let opts = BalanceOptions::default();
        let output = balance_command(&journal, &opts);
        assert_eq!(output, "");
    }

    #[test]
    fn balance_single_transaction() {
        let text = "\
2024/01/15 Grocery Store
    Expenses:Food       $42.50
    Assets:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions::default();
        let output = balance_command(&journal, &opts);

        assert!(output.contains("$42.50"));
        assert!(output.contains("$-42.50"));
        assert!(output.contains("Expenses:Food"));
        assert!(output.contains("Assets:Checking"));
        // Total should be zero
        assert!(output.contains("--------------------"));
        let lines: Vec<&str> = output.trim_end().split('\n').collect();
        let last_line = lines.last().unwrap().trim();
        assert_eq!(last_line, "0");
    }

    #[test]
    fn balance_multiple_transactions() {
        let text = "\
2024/01/01 Paycheck
    Assets:Bank:Checking     $1000.00
    Income:Salary

2024/01/02 Groceries
    Expenses:Food              $50.00
    Assets:Bank:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions::default();
        let output = balance_command(&journal, &opts);

        assert!(output.contains("$950.00"));
        assert!(output.contains("Expenses:Food"));
        assert!(output.contains("Income:Salary"));
    }

    #[test]
    fn balance_flat_mode() {
        let text = "\
2024/01/01 Paycheck
    Assets:Bank:Checking     $1000.00
    Income:Salary

2024/01/02 Groceries
    Expenses:Food              $50.00
    Assets:Bank:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions {
            flat: true,
            ..Default::default()
        };
        let output = balance_command(&journal, &opts);

        // Flat mode shows full account names
        assert!(output.contains("Assets:Bank:Checking"));
        assert!(output.contains("Expenses:Food"));
        assert!(output.contains("Income:Salary"));
    }

    #[test]
    fn balance_no_total() {
        let text = "\
2024/01/15 Grocery Store
    Expenses:Food       $42.50
    Assets:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions {
            no_total: true,
            ..Default::default()
        };
        let output = balance_command(&journal, &opts);

        assert!(!output.contains("----"));
    }

    #[test]
    fn balance_with_filter() {
        let text = "\
2024/01/01 Paycheck
    Assets:Bank:Checking     $1000.00
    Income:Salary

2024/01/02 Groceries
    Expenses:Food              $50.00
    Assets:Bank:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions {
            patterns: vec!["Expenses".to_string()],
            ..Default::default()
        };
        let output = balance_command(&journal, &opts);

        assert!(output.contains("Expenses"));
        assert!(!output.contains("Income"));
        assert!(!output.contains("Assets"));
    }

    #[test]
    fn balance_flat_with_filter() {
        let text = "\
2024/01/01 Paycheck
    Assets:Bank:Checking     $1000.00
    Income:Salary

2024/01/02 Groceries
    Expenses:Food              $50.00
    Assets:Bank:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions {
            flat: true,
            patterns: vec!["Assets".to_string()],
            ..Default::default()
        };
        let output = balance_command(&journal, &opts);

        assert!(output.contains("Assets:Bank:Checking"));
        assert!(!output.contains("Income"));
        assert!(!output.contains("Expenses"));
    }

    #[test]
    fn balance_hierarchical_collapsing() {
        // When an account has only one child and no direct postings,
        // it should be collapsed in tree mode.
        let text = "\
2024/01/15 Test
    Expenses:Food:Dining       $42.50
    Assets:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions::default();
        let output = balance_command(&journal, &opts);

        // "Expenses:Food:Dining" should be collapsed since Expenses and
        // Expenses:Food have no direct postings and only one child each.
        // The display should show "Expenses:Food:Dining" as a collapsed entry.
        assert!(output.contains("Dining"));
    }

    #[test]
    fn balance_depth_limit() {
        let text = "\
2024/01/01 Paycheck
    Assets:Bank:Checking     $1000.00
    Income:Salary

2024/01/02 Groceries
    Expenses:Food              $50.00
    Assets:Bank:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions {
            depth: 1,
            ..Default::default()
        };
        let output = balance_command(&journal, &opts);

        // At depth 1, should show only top-level accounts
        assert!(output.contains("Assets"));
        assert!(output.contains("Expenses"));
        assert!(output.contains("Income"));
        // Should NOT show nested accounts
        assert!(!output.contains("Bank:Checking"));
        assert!(!output.contains("Food"));
    }

    #[test]
    fn balance_separator_and_total() {
        let text = "\
2024/01/15 Transfer
    Assets:Savings       $500.00
    Assets:Checking     $-500.00
";
        let journal = parse_journal(text);
        let opts = BalanceOptions::default();
        let output = balance_command(&journal, &opts);

        // The total should be zero since it's a transfer within Assets
        assert!(output.contains("--------------------"));
        let lines: Vec<&str> = output.trim_end().split('\n').collect();
        let last = lines.last().unwrap().trim();
        assert_eq!(last, "0");
    }

    #[test]
    fn balance_amount_right_aligned() {
        let text = "\
2024/01/15 Test
    Expenses:Food       $42.50
    Assets:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions::default();
        let output = balance_command(&journal, &opts);

        // Check that amounts are right-aligned within AMOUNT_WIDTH
        for line in output.lines() {
            if line.starts_with('-') {
                continue; // separator
            }
            if line.contains("  ") && !line.trim().starts_with("0") {
                // Amount + account name line
                let amount_part = &line[..AMOUNT_WIDTH];
                // Amount should be right-aligned (starts with spaces)
                assert!(
                    amount_part.starts_with(' ') || amount_part.starts_with('$'),
                    "Amount not right-aligned: {:?}",
                    amount_part
                );
            }
        }
    }

    // ---- programmatic journal construction ---------------------------------

    #[test]
    fn balance_programmatic_journal() {
        let mut journal = Journal::new();

        let acct1 = journal.find_account("Expenses:Food", true).unwrap();
        let acct2 = journal.find_account("Assets:Checking", true).unwrap();

        let mut xact = Transaction::with_payee("Grocery Store");
        xact.add_post(Post::with_account_and_amount(
            acct1,
            Amount::parse("$42.50").unwrap(),
        ));
        xact.add_post(Post::with_account(acct2));
        journal.add_xact(xact).unwrap();

        let opts = BalanceOptions::default();
        let output = balance_command(&journal, &opts);

        assert!(output.contains("$42.50"));
        assert!(output.contains("$-42.50"));
    }

    #[test]
    fn balance_case_insensitive_filter() {
        let text = "\
2024/01/15 Test
    Expenses:Food       $42.50
    Assets:Checking
";
        let journal = parse_journal(text);
        let opts = BalanceOptions {
            flat: true,
            patterns: vec!["expenses".to_string()],
            ..Default::default()
        };
        let output = balance_command(&journal, &opts);
        assert!(output.contains("Expenses:Food"));
    }
}
